src/coreclr/vm/codeman.cpp

```diff
@@ -438,34 +443,26 @@ void UnwindInfoTable::AddToUnwindInfoTable(UnwindInfoTable** unwindInfoPtr, PT_R
 {
     STANDARD_VM_CONTRACT;
-    {
-        // CodeHeapIterator holds the m_CodeHeapCritSec, which ensures code heaps don't get deallocated while being walked
-        EEJitManager::CodeHeapIterator heapIterator(NULL);
-
-        // Currently m_CodeHeapCritSec is given the CRST_UNSAFE_ANYMODE flag which allows it to be taken in a GC_NOTRIGGER
-        // region but also disallows GC_TRIGGERS.  We need GC_TRIGGERS because we take another lock.   Ideally we would
-        // fix m_CodeHeapCritSec to not have the CRST_UNSAFE_ANYMODE flag, but I currently reached my threshold for fixing
-        // contracts.
-        CONTRACT_VIOLATION(GCViolation);
 
-        while(heapIterator.Next())
+    // CodeHeapIterator ensures code heaps don't get deallocated while being walked
+    CodeHeapIterator heapIterator = ExecutionManager::GetEEJitManager()->GetCodeHeapIterator();
+    while(heapIterator.Next())
+    {
+        MethodDesc *pMD = heapIterator.GetMethod();
+        if(pMD)
         {
-            MethodDesc *pMD = heapIterator.GetMethod();
-            if(pMD)
+            PCODE methodEntry =(PCODE) heapIterator.GetMethodCode();
+            RangeSection * pRS = ExecutionManager::FindCodeRange(methodEntry, ExecutionManager::GetScanFlags());
+            _ASSERTE(pRS != NULL);
+            _ASSERTE(pRS->_pjit->GetCodeType() == (miManaged | miIL));
+            if (pRS != NULL && pRS->_pjit->GetCodeType() == (miManaged | miIL))
             {
-                PCODE methodEntry =(PCODE) heapIterator.GetMethodCode();
-                RangeSection * pRS = ExecutionManager::FindCodeRange(methodEntry, ExecutionManager::GetScanFlags());
-                _ASSERTE(pRS != NULL);
-                _ASSERTE(pRS->_pjit->GetCodeType() == (miManaged | miIL));
-                if (pRS != NULL && pRS->_pjit->GetCodeType() == (miManaged | miIL))
-                {
-                    // This cast is justified because only EEJitManager's have the code type above.
-                    EEJitManager* pJitMgr = (EEJitManager*)(pRS->_pjit);
-                    CodeHeader * pHeader = pJitMgr->GetCodeHeaderFromStartAddress(methodEntry);
-                    int unwindInfoCount = pHeader->GetNumberOfUnwindInfos();
-                    for(int i = 0; i < unwindInfoCount; i++)
-                        AddToUnwindInfoTable(&pRS->_pUnwindInfoTable, pHeader->GetUnwindInfo(i), pRS->_range.RangeStart(), pRS->_range.RangeEndOpen());
-                }
+                // This cast is justified because only EEJitManager's have the code type above.
+                EEJitManager* pJitMgr = (EEJitManager*)(pRS->_pjit);
+                CodeHeader * pHeader = pJitMgr->GetCodeHeaderFromStartAddress(methodEntry);
+                int unwindInfoCount = pHeader->GetNumberOfUnwindInfos();
+                for(int i = 0; i < unwindInfoCount; i++)
+                    AddToUnwindInfoTable(&pRS->_pUnwindInfoTable, pHeader->GetUnwindInfo(i), pRS->_range.RangeStart(), pRS->_range.RangeEndOpen());
```

Comment:
Can we end up seeing a method that was created after the iteration started and end up publishing the unwind info again? Is the system going to handle the republishing gracefully?

One possible way to handle this is to publish these unwind infos unconditionally all the time. They are likely published most of the time anyway since there is typically some tracing session enable, so we are not saving much by publishing them on demand.

Reply:
> Can we end up seeing a method that was created after the iteration started and end up publishing the unwind info again? Is the system going to handle the republishing gracefully?

I don't think so. The iterator here captures a snapshot of the current range and only examines that range of data. See `CodeHeapIterator` constructor in `codeman.cpp` for where I'm capturing the current state. Perhaps I'm missing something here.

Reply:
I am worried about the following sequence:

- Iteration starts
- Dynamic methods get created and reuses space in one of the existing heaps
- Unwind info for this dynamic method gets published here
- Iteration reaches the heap where dynamic method reused space in step 2
- The unwind gets published again

Is there something that prevents this sequence from happening?

Reply:
As far as I can tell, https://github.com/dotnet/runtime/pull/116354/files#diff-eba17a72b6caa8b52b196f3b84bfd20dfeaf39a89f410943a692d1f33f569f3bR536 does not prevent reuse of the free space that's within the range for dynamic methods

Reply:
Nevermind. I see what you are suggesting here. I assumed this function was doing something else.

Reply:
Done.